From 4c32050d9e807f25501e14be3c50a2036aaa1fa6 Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Tue, 25 Nov 2014 16:34:04 -0500
Subject: [fs] btrfs: add RENAME_NOREPLACE

Message-id: <20141125163404.18115.89381.stgit@warthog.procyon.org.uk>
Patchwork-id: 100915
O-Subject: [PATCH 16/47] btrfs: add RENAME_NOREPLACE
Bugzilla: 985875
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Miklos Szeredi <mszeredi@suse.cz>

RENAME_NOREPLACE is trivial to implement for most filesystems: switch over
to ->rename2() and check for the supported flags.  The rest is done by the
VFS.

Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Cc: Chris Mason <clm@fb.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Original-commit: 80ace85c915d0f41016f82917218997b72431258
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index db05561..aa66902 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -8482,6 +8482,16 @@ out_notrans:
  return ret;
 }
 
+static int btrfs_rename2(struct inode *old_dir, struct dentry *old_dentry,
+    struct inode *new_dir, struct dentry *new_dentry,
+    unsigned int flags)
+{
+ if (flags & ~RENAME_NOREPLACE)
+  return -EINVAL;
+
+ return btrfs_rename(old_dir, old_dentry, new_dir, new_dentry);
+}
+
 static void btrfs_run_delalloc_work(struct btrfs_work *work)
 {
  struct btrfs_delalloc_work *delalloc_work;
@@ -8965,7 +8975,7 @@ static const struct inode_operations btrfs_dir_inode_operations = {
  .link  = btrfs_link,
  .mkdir  = btrfs_mkdir,
  .rmdir  = btrfs_rmdir,
- .rename  = btrfs_rename,
+ .rename2 = btrfs_rename2,
  .symlink = btrfs_symlink,
  .setattr = btrfs_setattr,
  .mknod  = btrfs_mknod,
-- 
1.7.1